### DISCUSSION TITLE
Need Advice - what is the recommended way to opptimize & test Fingerprint

### DISCUSSION THREAD

**Discussion by Mozila4060** [fingerprintjs/fingerprintjs#676] [General]
HI,

We have installed FPJS on our site that allows capturing FP information and analyze the performance of each Attribute. Each user we can identify with the phone number and therefore can measure the uniqueness of the FP and monitor the changes in the attributes. 

We start analyzing the data and see cases when for the same user/browser some values are changing in very short period of time. (10-20 minutes).

Looking for advice on how to test and optimize the performance. What optimization steps to take? 

For instance, we noticed that most attribute that changes are:
- languages
- fontPreferences
- canvas (text value)

in **languages** changes looks like:
**from** [["ru"],["ru"]]  **to ->**	[["ru-RU"],["ru-RU","ru","en-US","en"]]
**from** [["ru"]]	**to ->** [["ru-RU"]]

in **fontPreferences** changes looks like:
**from** {default=167, apple=167, serif=167, sans=144, mono=136, min=15, system=144}	**- interger values**
**to** {default=164.71875, apple=164.71875, serif=164.71875, sans=145.90625, mono=132.625, min=10.296875, system=145.90625}

Other cases they just changed:
**from** {default=149.3125, apple=129.923828125, serif=149.3125, sans=144.0078125, mono=132.62158203125, min=9.33203125, system=146.09375}	
**to** {default=149.3125, apple=153.53125, serif=149.3125, sans=144.0078125, mono=132.62158203125, min=9.33203125, system=146.09375}

What does it mean? This is not stable across browsers or there is other reasons? How can we stabilize the result of web fingreprint?

**Comment by makma** [comment]
Hello @Mozila4060!
From your descriptions and traits, it seems the same users have been using different browsers. It's a known behavior that e.g. Chrome returns a different `languages` format than firefox. The same applies to `fontPreferences` with rounding and returned values.
Could you please confirm whether the given properties are collected from the same or different browsers? - at first, I'd take a look at user agents. If so I believe this is a proper behavior as these signals are not meant to be stable amongst different browsers even for the same users.

> How can we stabilize the result of web fingerprint?

- I'm afraid your expectations are not set clearly - FingerprintJS is a tool for browser fingerprinting not for identifying devices or "web access" per se.

> This is not stable across browsers 

- No, these properties are not meant to be stable amongst browsers.

**Reply by Mozila4060** [reply]
@makma  Thank you. 
1. Will ask the developer to double-check if the browser is the same or different for recorded library loads. 
2. I understand what FingerprintJS is. My request is help to find a way to configure FingerprintJS to producing stable results over a period of time and learn the extent of the stability. The duration of the "period" we will measure. I understand the FP will change eventually. 
3. We would like to learn the stability in instability per attribute.

**Reply by makma** [reply]
Hello @Mozila4060! I've just filed the[ issue for collecting info about stable and independent components](https://github.com/fingerprintjs/fingerprintjs/issues/677). Feel free to add any related info and your observation about stability. Once we have enough data we will distill it and create documentation pages.
